File: 2025/day-04/part2.py
import sys

if len(sys.argv) > 1:
	INPUT = sys.argv[1]
else:
	INPUT = "input1.txt"

with open(INPUT) as f:
	lines = f.readlines()
	roll_positions = set()
	for y, line in enumerate(lines):
		for x, c in enumerate(line):
			if c == "@":
				roll_positions.add((x, y))
	print(len(roll_positions))
	removable_rolls = 0
	while True:
		removed = set()
		for x, y in roll_positions:
			count = 0
			if (x - 1, y) in roll_positions:
				count += 1
			if (x - 1, y + 1) in roll_positions:
				count += 1
			if (x - 1, y - 1) in roll_positions:
				count += 1
			if (x, y - 1) in roll_positions:
				count += 1
			if (x, y + 1) in roll_positions:
				count += 1
			if (x + 1, y) in roll_positions:
				count += 1
			if (x + 1, y + 1) in roll_positions:
				count += 1
			if (x + 1, y - 1) in roll_positions:
				count += 1
			# not too many neighbors: can take roll
			if count < 4:
				removed.add((x, y))
		if len(removed) == 0:
			break
		removable_rolls += len(removed)
		for x, y in removed:
			if (x, y) in roll_positions:
				roll_positions.remove((x, y))
	print(removable_rolls)

File: 2025/day-04/src/part2.rs
use std::collections::HashSet;

use glam::IVec2;
const DIRECTIONS: [IVec2; 8] = [
	IVec2::NEG_X,
	IVec2::NEG_Y,
	IVec2::X,
	IVec2::Y,
	IVec2::ONE,
	IVec2::NEG_ONE,
	IVec2::new(1, -1),
	IVec2::new(-1, 1),
];

#[tracing::instrument]
pub fn process(_input: &str) -> miette::Result<String> {
	let mut roll_positions = _input
		.lines()
		.enumerate()
		.flat_map(|(y, line)| {
			line.chars().enumerate().filter_map(move |(x, c)| {
				if c == '@' {
					// dbg!(&x, &y);
					Some(IVec2::new(x as i32, y as i32))
				} else {
					None
				}
			})
		})
		.collect::<HashSet<IVec2>>();
	let mut removed_total = 0;
	loop {
		let remainder: HashSet<IVec2> = roll_positions
			.iter()
			.filter(|&roll| {
				// keep those who can't be taken: inverse part1
				DIRECTIONS
					.iter()
					.filter(|&dir_vec| roll_positions.contains(&(roll + dir_vec)))
					.count() >= 4
			})
			.cloned()
			.collect();
		if remainder.len() == roll_positions.len() {
			break;
		} else {
			removed_total += roll_positions.len() - remainder.len();
			roll_positions = remainder;
		}
	}
	Ok(removed_total.to_string())
}
#[cfg(test)]
mod tests {
	use super::*;

	#[test]
	fn test_process() -> miette::Result<()> {
		let input = "..@@.@@@@.
@@@.@.@.@@
@@@@@.@.@@
@.@@@@..@.
@@.@@@@.@@
.@@@@@@@.@
.@.@.@.@@@
@.@@@.@@@@
.@@@@@@@@.
@.@.@@@.@.";
		assert_eq!("43", process(input)?);
		Ok(())
	}
}
